File: runtime/java/units/rt/trap.java
// A wasm trap: a runtime fault (out-of-bounds, integer overflow, ...). `trap`
// is void and throws, so a generated `Rt.trap(...)` statement needs no Java
// `throw` at the call site — which avoids an "unreachable statement" error
// after it (ADR-30). Recovered at the public boundary.
static final class Trap extends RuntimeException {
    Trap(String msg) {
        super(msg);
    }
}

static void trap(String msg) {
    throw new Trap(msg);
}

File: runtime/go/units/rt/trap.go
// A wasm trap: a runtime fault (out-of-bounds, integer overflow, ...) raised
// with panic and recovered at the public boundary, mirroring Ruby/Python's
// Rt.Trap.
type rtTrap struct{ msg string }

func (e *rtTrap) Error() string { return e.msg }

func (rt) trap(msg string) {
    panic(&rtTrap{msg})
}
